README.md: Add details on how the tests are run (#20)
samba-in-kubernetes/sit-test-cases: README.md

# Samba integration tests

The tests are part of the Samba Integrated testing project
[sit-environment](https://github.com/samba-in-kubernetes/sit-environment)
and are primarily used to test Samba servers exporting a clustered

> filesystem such as cephfs, glusterfs and others.

Reviewer: My bad. Withdrawing the above suggestion.

### Prerequisite:
The tests are designed to run against any SMB server. The only
prerequisite is that we have a SMB server exporting a filesystem which
can be accessed over the network.
Refer:
[server_setup](https://wiki.samba.org/index.php/Setting_up_Samba_as_a_Standalone_Server)

> 

Author: If there is any doc about exporting the filesystem, we can add it too here.
@spuiuk Are you aware of any such relevant doc?

[sit-environment](https://github.com/samba-in-kubernetes/sit-environment)
and are primarily used to test Samba servers exporting a clustered
filesystem such as cephfs, glusterfs and others.

### Prerequisite:
The tests are designed to run against any SMB server. The only
prerequisite is that we have a SMB server exporting a filesystem which
can be accessed over the network.
Refer:
[server_setup](https://wiki.samba.org/index.php/Setting_up_Samba_as_a_Standalone_Server)
[filesytem_export_with_glusterfs](https://wiki.samba.org/index.php/Samba_CTDB_GlusterFS_Cluster_HowTo)

### Quick guide on running the test-cases:
- Install sit-cases
  ```
     $ git clone https://github.com/samba-in-kubernetes/sit-test-cases.git
  ```

- Rename test-info.yml.example file under sit-test-cases to test-info.yml
  ```
      $ cd sit-test-cases
      $ mv test-info.example test-info.yml
  ```

- Update the relevant information in test-info.yml such as private and
  public interfaces, exported share names and test users and passwords from the
  previously set up samba server.

- Run the required tests from the root of the sit-test-cases repo:
  Example:
  ```

>       #PYTHONPATH=`pwd` TESTINFO_FILE=test-info.yml pytest -v testcases/smbtorture

Reviewer 2: a space between the '#' and PYTHONPATH please. 

also, can you please mention that some tests require you to mount the filesystem using the cifs kernel module and you may need to run the tests as root.

It looks fine otherwise.

filesystem such as cephfs, glusterfs and others.

### Prerequisite:
The tests are designed to run against any SMB server. The only
prerequisite is that we have a SMB server exporting a filesystem which
can be accessed over the network.
Refer:
[server_setup](https://wiki.samba.org/index.php/Setting_up_Samba_as_a_Standalone_Server)
[filesytem_export_with_glusterfs](https://wiki.samba.org/index.php/Samba_CTDB_GlusterFS_Cluster_HowTo)

### Quick guide on running the test-cases:
- Install sit-cases
  ```
     $ git clone https://github.com/samba-in-kubernetes/sit-test-cases.git
  ```

- Rename test-info.yml.example file under sit-test-cases to test-info.yml
  ```
      $ cd sit-test-cases
      $ mv test-info.example test-info.yml
  ```

- Update the relevant information in test-info.yml such as private and
  public interfaces, exported share names and test users and passwords from the
  previously set up samba server.

- Run the required tests from the root of the sit-test-cases repo:
  Example:
  ```
      # PYTHONPATH=`pwd` TESTINFO_FILE=test-info.yml pytest -v testcases/smbtorture
  ```

NOTE:
- The tests are to be run as root.

> - Mount the filesystem using cifs kernal module to run the tests that use cifs

Reviewer: ```suggestion
- Mount the filesystem using cifs kernel module to run the tests that use cifs
```

and are primarily used to test Samba servers exporting a clustered
filesystem such as cephfs, glusterfs and others.

### Prerequisite:
The tests are designed to run against any SMB server. The only
prerequisite is that we have a SMB server exporting a filesystem which
can be accessed over the network.
Refer:
[server_setup](https://wiki.samba.org/index.php/Setting_up_Samba_as_a_Standalone_Server)
[filesytem_export_with_glusterfs](https://wiki.samba.org/index.php/Samba_CTDB_GlusterFS_Cluster_HowTo)

### Quick guide on running the test-cases:
- Install sit-cases
  ```
     $ git clone https://github.com/samba-in-kubernetes/sit-test-cases.git
  ```

- Rename test-info.yml.example file under sit-test-cases to test-info.yml
  ```
      $ cd sit-test-cases
      $ mv test-info.example test-info.yml
  ```

- Update the relevant information in test-info.yml such as private and
  public interfaces, exported share names and test users and passwords from the
  previously set up samba server.

- Run make test to confirm the sanity of the setup before running the tests
  (Can be used whenever new set of changes are introduced in sit-test-cases):

>   $ make test

Reviewer 2: make test run the entire test suite. I would not use this as a sanity test
make sanity_test runs just the sanity test to confirm that the SMB server is working as expected and ready for the full test suite.

The individual tests below is just to run specific tests and  should not be the default step to run. We only run that when we are testing to fix an individual test case.

# Samba integration tests

The tests are part of the Samba Integrated testing project
[sit-environment](https://github.com/samba-in-kubernetes/sit-environment)
and are primarily used to test Samba servers exporting a clustered
filesystem such as cephfs, glusterfs and others.

### Prerequisite:
The tests are designed to run against any SMB server. The only
prerequisite is that we have a SMB server exporting a filesystem which
can be accessed over the network.
Refer:
[server_setup](https://wiki.samba.org/index.php/Setting_up_Samba_as_a_Standalone_Server)
[filesytem_export_with_glusterfs](https://wiki.samba.org/index.php/Samba_CTDB_GlusterFS_Cluster_HowTo)

### Quick guide on running the test-cases:

> - Install sit-cases

Reviewer 3: Install sit-test-cases

  ```
      $ git clone https://github.com/samba-in-kubernetes/sit-test-cases.git
  ```

- Rename test-info.yml.example file under sit-test-cases to test-info.yml
  ```
      $ cd sit-test-cases
      $ mv test-info.example test-info.yml
  ```

- Update the relevant information in test-info.yml such as private and
  public interfaces, exported share names and test users and passwords from the
  previously set up samba server.

- Run make test to confirm the sanity of the setup before running the tests
  (Can be used whenever new set of changes are introduced in sit-test-cases):
  ```

>       $ make test

Reviewer 2: make test runs the full test suite. This is not a sanity test

make sanity_test runs the actual sanity test which confirms that the smb server can be accessed and is ready to perform the full test on.

  ```

- Run the required tests from the root of the sit-test-cases repo:
  Example:
  ```

>       # PYTHONPATH=`pwd` TESTINFO_FILE=test-info.yml pytest -v testcases/smbtorture

Reviewer 2: I only use this when I need to run a specific test. This is usually when I am working on fixing a particular problem. So this is an exception and shouldn't be needed as a rule. This however still needs to be documented in case the user wants to run a particular test.

  ```

NOTE:

> - The tests are to be run as root.

Reviewer 3: Tests must be run as root

> - Mount the filesystem using cifs kernel module to run the tests that use cifs

Reviewer 2: would rewrite this part.

Some tests are performed against a share mounted using the cifs kernel module. This particular action requires root access.

  share/mount.

>   test_mount.py and test_consistency.py are few such tests that use cifs share.

Reviewer 2: test_consistency uses smbclient to connect to the server to perform a sanity test. This doesn't actually use the cifs  mount.

# Samba integration tests

This repository contains tests for Samba when used with clustered file-systems
such as cephfs, glusterfs or other Software-Defined Storage systems.

The test environment can either be created using
[sit-environment](https://github.com/samba-in-kubernetes/sit-environment) or by
manual steps with minimal resources.

## Quick guide on environment creation and running tests manually:

### Prerequisites:
- Install Samba using [download guide](https://www.samba.org/samba/download/)
  and [setup guide](https://www.samba.org/samba/download/)
- Set up [CTDB and Clustered Samba](https://wiki.samba.org/index.php/CTDB_and_Clustered_Samba)

> 

Reviewer 2: We don't really need a CTDB with clustered samba installed. I recommend that we change it to something similar to
```
The tests are part of the Samba Integrated testing project which is primarily used to test Samba servers exporting a clustered filesystem. 

The tests itself are designed to be run against any SMB server. The only prerequisites are that we have a SMB server exporting a filesystem which can be accessed over the network.

```

### Run test-cases:
- Install sit-cases on samba client.
  ```
     $ git clone https://github.com/samba-in-kubernetes/sit-test-cases.git
  ```

- Rename test-info.yml.example file under sit-test-cases to test-info.yml
  ```
      $ cd sit-test-cases
      $ mv test-info.example test-info.yml
  ```

- Update the relevant information such as private and public interfaces,
  exported share names and test users and passwords from the previously
  created clusterd samba setup inside test-info.yml

> - Export essential PATH variables inorder to access pytest and test-info.yml:

Reviewer 2: We can simply use the Makefile to do all the testing. 

The targets we are interested in are 
test - which runs the complete test suite
and sanity_test - a quick test to confirm that the smb server is accessible.

To run individual tests, I would simply use the command in the following manner from the root of the sit-test-cases repo
```
PYTHONPATH=`pwd` TESTINFO_FILE=test-info.yml pytest -v testcases/smbtorture
```

### Prerequisite:
The tests are designed to run against any SMB server. The only
prerequisite is that we have a SMB server exporting a filesystem which
can be accessed over the network.
Refer:
[server_setup](https://wiki.samba.org/index.php/Setting_up_Samba_as_a_Standalone_Server)
[filesytem_export_with_glusterfs](https://wiki.samba.org/index.php/Samba_CTDB_GlusterFS_Cluster_HowTo)

### Quick guide on running the test-cases:
- Install sit-test-cases
  ```
      $ git clone https://github.com/samba-in-kubernetes/sit-test-cases.git
  ```

- Rename test-info.yml.example file under sit-test-cases to test-info.yml
  ```
      $ cd sit-test-cases
      $ mv test-info.example test-info.yml
  ```

- Update the relevant information in test-info.yml such as private and
  public interfaces, exported share names and test users and passwords from the
  previously set up samba server.

- Run make sanity_test to confirm the sanity of the setup before running the tests
  (Can be used whenever new set of changes are introduced in sit-test-cases):
  ```
      $ make sanity_test
  ```

- Run full test suite using make test:
  ```
      $ make test
  ```

> - Run an indivisual test case (can be used for debugging):

Reviewer 2: spelling. "individual". Also please can you specify that the command has to be run in the root of the repo. The PYTHONPATH=`pwd` sets the search path of the library to the root of the repo where we can find the testhelper folder.

# Samba integration tests

> This repo contains tests for Samba and storage integration that are run on the client machines.

Reviewer 4: Perhaps we can be a bit clearer here what storage integration means. Something like: "This repo contains tests for Samba when used with Clustered or other Software-Defined Storage systems"... ?

> This repo contains tests for Samba when used with Clustered or other Software-Defined Storage systems.

Reviewer 3: Consider changing to: "when used with clustered file-system, such as fephfs, glusterfs or other".

> [sit-environment](https://github.com/samba-in-kubernetes/sit-environment) creates an environment on which these tests can be run using vagrant by setting up

Reviewer 3: In most README.md files we try to limit to 80-chars lines. Some of use still `vim` on terminal with large fonts...

Author: I had now limited all the lines to 80 charecters, except for the link in line 21, which on splitting does not behave as link anymore in .md @synarete

virtual machines with the desired backend storage, ctdb and samba installed.
Ansible playbooks in sit-environment run the tests in [sit-test-cases](https://github.com/samba-in-kubernetes/sit-test-cases).

> ## We can also manually run the individual tests in sit-environment on our local machines using the following steps:

Reviewer 4: I'm being nitpicky, but this seems a bit long for a title.

> PRE-REQUISITES:

Reviewer 3: Why not: ```### Prerequisites:```

- [sit-environment](https://github.com/samba-in-kubernetes/sit-environment) should be locally intalled.

> - After installation, sit-envornment should be setup using [setup guide](https://github.com/samba-in-kubernetes/sit-environment/blob/main/docs/Fedora-Setup.md)

Reviewer 4: Typo: `sit-envornment`

When used as a verb it is more grammatical to spell "set up" as two words. It's typically spelled as one word only when you use it as a noun.

> STEPS TO RUN INDIVIDUAL SIT TEST CASES:

Reviewer 3: Why not: ```### Run individual test-cases:```

- check the vagrant status to know the status of machines created by the sit-environment
  ```
      # pwd
        /home/<user>/sit-environment
      # cd vagrant
      # vagrant status
        Current machine states:

        setup0                    running (libvirt)
        client0                   running (libvirt)
        storage0                  running (libvirt)
        storage1                  running (libvirt)
  ```

> - Login to client0 inorder the manually run the tests on the client

Reviewer 4: More nitpicks: "Log in" and "in order to"

  ```
      # vagrant ssh client0
  ```
- Change to root user
  ```

>       # sudo -i

Reviewer 4: Why does this need to run as root?  If we can' avoid it (for now), can we avoid using a root shell and just run `sudo <my particular command>` instead?

Author: I have updated this PR as per all other suggestions of yours, expect this thread. Adding other reviewers as well, so that we can get collective opinion on this. Please pitch in if you have any views on my last comment @phlogistonjohn

Reviewer 4: It's fine for now. You're trying to document how it works not how it should be after all.

# Samba integration tests

> This branch contains the actual tests for the cluster nodes and are run on the client machines.

Reviewer 5: branch --> repository

cephfs, glusterfs or other Software-Defined Storage systems.

[sit-environment](https://github.com/samba-in-kubernetes/sit-environment)

> creates an environment on which these tests can be run. Environment is created

Reviewer 5: ```suggestion
creates an environment on which these tests can be run.  The environment is created
```

> using vagrant by setting up virtual machines with the desired backend storage,

Reviewer 5: ```suggestion
on  virtual machines that are set up  by vagrant and then configured with the desired backend storage,
```

Reviewer 5: I do in fact doubt that this level of detail is right in the  test-cases repo docs: it is information that should be in the sit-environment docs.

Author: @obnoxxx done!
